### PR TITLE
Added rrp_ip host variable to support redundant corosync rings setup

### PR DESCRIPTION
**Pull request by SpitchAG**
currently the role doesnt support redundant ring creation,
we fixed it by adding a rrp_ip host  variable in the play .hosts file for each node  e.g.: 
[all]
ip-172-31-1-9  rrp_ip=172.31.5.72 #must be valid ip, no name mapping is created in /etc/hosts file  
ip-172-31-1-197  rrp_ip=172.31.5.112

As you see In the hosts file, for each host/node defined, if a rrp_ip variable is defined (alternative ip for this node) a redundant ring will be created.
    (each node/host must have an alt ip defined)

maybe there is better way to do it, but i didnt see easy way

### CODE REVIEW

**Comment by SpitchAG** [comment]
maybe it is better to use rrp_interface instead of rrp_ip, and use ansible_interface fact to find out ipv4 ?
or support both,

**Comment by OndrejHome** [comment]
Hello @SpitchAG ,
Thank you for using the ha-cluster-pacemaker role and for opening up this PR.
You have a good point that while `pcs_cluster` module supports RRP this role doesn't use that now.

Sorry for a bit longer text here where I try to explain the situation around RRP for this role.

As for your last comment I would incline to use of `rrp_interface` over `rrp_ip`. The "valid" use cases for `rrp_ip` that I can imagine are:
- using different IP on same interface as primary ring_0 - unless there is special routing that would route that IP other way that other heartbeat I don't think this can bring benefit in stability, but can be used for "testing cluster" where you have only one interface, but you want to use/test RRP - I would still prefer not encouraging this approach
- using non-primary IP on some interface - while interfaces can have one primary IP, they can have more secondary IPs on same interface - unless for deliberately knowing that you don't want to use the primary IP on given interface I would consider this a rather corner case

I would consider this role to aim on people that want to "get cluster running" or "get it running with RRP" in simple way. For that reason I think for RRP support we should focus on `rrp_interface` which can be used to encourage users to use different interface from main ring_0. To further strengthen that a small check could be added (NOTE: code only for idea, syntax might be incorrect:
~~~
- name: check if RRP interface is different from ring_0 cluster interface
  fail:
    msg: rrp_interface must be different from main cluster interface
  when: 
    ## basic check that we have not selected manually same interfaces for both MAIN link and RRP link
    (rrp_interface is defined and cluster_net_iface is defined and cluster_net_iface == rrp_interface) 
    or 
    ## if only rrp_interface was specified, lets make sure it is not the default interface which would mean same IP
    (rrp_interface is defined and cluster_net_iface is not defined and hostvars[item]['ansible_'+rrp_interface]['ipv4']['address'] == hostvars[item]['ansible_default_ipv4']['address']) 
    or
    ## sanity check that selected rrp_interface actually exists, maybe checking even further that it has IP would be good idea
    (rrp_interface is defined and hostvars[item]['ansible_'+rrp_interface] is not defined) 
~~~

 We can figure out the IPv4 from interface using the `hostvars[item]['ansible_'+rrp_interface]['ipv4']['address']` variable. (this is very similar to approach using alternative interface that figures out the IP of that interface for example here - https://github.com/OndrejHome/ansible.ha-cluster-pacemaker/blob/master/tasks/main.yml#L50 )

And with that it should be possible to generate the hostname into `/etc/hosts`. Example below will generate 'hostname-rrp' hostnames with `rrp_interface` IPv4.
~~~
- name: Add hosts to /etc/hosts (for RRP interface)
  lineinfile:
    dest: '/etc/hosts'
    regexp: "{{ hostvars[item]['ansible_'+rrp_interface]['ipv4']['address'] }} "
    line: >
      {{ hostvars[item]['ansible_'+rrp_interface]['ipv4']['address'] }}
      {{ hostvars[item]['ansible_fqdn'].split('.')[0] }}-rrp
  with_items: "{{ play_hosts }}"
  when: rrp_interface is defined and hostvars[item]['ansible_'+rrp_interface] is defined

~~~

One concern I have is what would happen if the `rrp_interface gets specified only for some nodes, but not all.

Most importantly for getting this change merged some documentation will be needed. When `rrp_interface` will be used then at least mentioning it in `README.md` and `defaults/main.yml` should do the trick.

---

In any case I think the use case for including RRP into this role makes sense. Please let me know if you agree with above feedback or have comments to it AND if you would be able to incorporate it into this PR so it can be merged.

**Comment by SpitchAG** [comment]
Hi,

agree that only relevant case is having rrp on 2 different interfaces, using primary address on both.
agree having extra checks, i can try to add those,

However not 100% sure about the use of rrp_interface instead of rrp_ip: corosync only needs an alternate ip to be dual, using an interface may lead to confusion (e.g. multiple ip on this interface ... even if at the end we force the use of the primary one).

Also I don't think it is good idea to update hosts, because you would have conflicts (same hostname pointing to 2 ips) and you may end up having the first ring faulty if there is no route between the 2 networks (because corosync will try to reach the other node hostname which get translated to ip of the alternate remote interface). At least  on my setup that is what happens.

**Comment by OndrejHome** [comment]
Hi SpitchAG,
> However not 100% sure about the use of rrp_interface instead of rrp_ip: corosync only needs an alternate ip to be dual,
- yes, but using 2 different IPs from same interface might be in most cases a mistake (unless you are testing or have doing RRP with IPs routed differently later on) - I believe that majority use cases would use RRP on different interface than ring_0 address. That is why I would rather encourage rrp_interface over rrp_ip. However I'm curious what is your use case where `rrp_ip` will be needed instead of just `rrp_interface`. Please let me know. (I'm trying to avoid `rrp_ip` so there is no need to also introduce `cluster_net_ip` variable :) - that is why there is only `cluster_net_iface` - having `rrp_ip` would make it logical to also have `cluster_net_ip` when I think about it)

> because you would have conflicts (same hostname pointing to 2 ips)
Looking at code I posted it seems to not work properly, please see the (tested example below)
~~~
---
- hosts: all
  remote_user: root
  tasks:
    # ring_0 hostname
    - name: Add hosts to /etc/hosts (using default (GW) IPv4 address)
      lineinfile:
        dest: '/etc/hosts'
        regexp: "{{ hostvars[item]['ansible_default_ipv4']['address'] }} "
        line: >
          {{ hostvars[item]['ansible_default_ipv4']['address'] }}
          {{ hostvars[item]['ansible_fqdn'].split('.')[0] }}
      with_items: "{{ play_hosts }}"
    # RRP hostname
    - name: Add hosts to /etc/hosts (for RRP interface)
      lineinfile:
        dest: '/etc/hosts'
        regexp: "{{ hostvars[item]['ansible_'+hostvars[item]['rrp_interface']]['ipv4']['address'] }} "
        line: >
          {{ hostvars[item]['ansible_'+hostvars[item]['rrp_interface']]['ipv4']['address'] }}
          {{ hostvars[item]['ansible_fqdn'].split('.')[0] }}-rrp
      with_items: "{{ play_hosts }}"
      when: rrp_interface is defined and hostvars[item]['ansible_'+rrp_interface] is defined
~~~
Above should generate `/etc/hosts` hostnames in following way:
~~~
10.0.0.1 host-1
10.0.0.2 host-2
192.168.0.1 host-1-rrp
192.168.0.2 host-2-rrp
~~~
Below example from my testing cluster
~~~
### node1 - fastvm-centos-7-8-71
# ip r
default via 192.168.22.1 dev eth0 proto dhcp metric 100
192.168.0.0/16 dev eth1 proto kernel scope link src 192.168.200.71 metric 101
192.168.22.0/24 dev eth0 proto kernel scope link src 192.168.22.71 metric 100
# ip a
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet 192.168.22.71/24 brd 192.168.22.255 scope global noprefixroute dynamic eth0
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet 192.168.200.71/16 brd 192.168.255.255 scope global noprefixroute eth1
# cat /etc/hosts
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
192.168.22.71 fastvm-centos-7-8-71
192.168.22.72 fastvm-centos-7-8-72
192.168.200.71 fastvm-centos-7-8-71-rrp
192.168.200.72 fastvm-centos-7-8-72-rrp

### node2 - fastvm-centos-7-8-72
# ip r
default via 192.168.22.1 dev eth1 proto dhcp metric 101
192.168.0.0/16 dev eth0 proto kernel scope link src 192.168.200.72 metric 100
192.168.22.0/24 dev eth1 proto kernel scope link src 192.168.22.72 metric 101
# ip a
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet 192.168.200.72/16 brd 192.168.255.255 scope global noprefixroute eth0
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet 192.168.22.72/24 brd 192.168.22.255 scope global noprefixroute dynamic eth1
# cat /etc/hosts
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
192.168.22.71 fastvm-centos-7-8-71
192.168.22.72 fastvm-centos-7-8-72
192.168.200.71 fastvm-centos-7-8-71-rrp
192.168.200.72 fastvm-centos-7-8-72-rrp
------
cat ansible-hosts
[cluster]
192.168.22.71 vm_name=fastvm-centos-7.8-71 rrp_interface=eth1
192.168.22.72 vm_name=fastvm-centos-7.8-72 rrp_interface=eth0
~~~
Does above make sense on how I thought about `/etc/hosts` hostnames?

**Comment by SpitchAG** [comment]
Hi
I agree that only relevant use case is having the 2 rings on different nic,
(please use what u think is best for your role)

As for the hosts I dont really get it:  in my case i dont need extra host names, i just need an alternate ip. the host is only needed for the primary ring which is reached through short host names. Adding new rrp hosts is no arm but i dont see the plus value of adding them to be honnest,

cheers

**Comment by OndrejHome** [comment]
Hi,

Thanks for reply. In regards to hostnames I'm biased as I have seen many RRP IPs to have hostname assigned so I though it is something really needed. Testing for myself on CentOS 7.8/8.2 I can see that it really is not required.

To put together acceptation criteria for this PR:
- [x] (required, already present in PR) edit that adds RRP IP to `pcs_cluster` module
- [ ] (required) implement check to verify that preferred RRP address is different from ring_0 address
- [ ] (required) documentation in `README.md` and `defaults/main.yml` describing new variable `rrp_interface`
- [ ] (nice-to-have) testing - letting me know on which OS versions this was tested would be nice to have. I plan checking this on latest version on CentOS (6.10, 7.8, 8.2) once ready.
- [ ] (optional) `rrp_ip` as additional alternative to `rrp_interface` if you really find that `rrp_interface` will not be enough for your use case

**Important question:**
  Would you like to adjust this PR to above points so I can merge it directly or you would prefer to leave it to me to merge current PR and do the points myself?

I will have some time during this week (from Thursday) to either provide feedback or deal with testing/coding/merging changes.

Thank you and have a good day!

---
RRP bug for `pcs_cluster` on CentOS 8.2: https://github.com/OndrejHome/ansible.pcs-modules-2/issues/33

**Comment by SpitchAG** [comment]
hello
i can try to add the required item and let u know os tested but it may take some times as i am working on some other stuff,
thanks

**Comment by OndrejHome** [comment]
Hi @SpitchAG,
In advance thank you once you find time to work on this. In such case just left me a note here (something like "I plan to look at this next week/month"). If I find time sooner than you then I will leave you the same note. 

As of now I have managed to get a proper RRP support to `pcs_cluster` module for both EL7 2 link RRP and also EL8 up to 8 redundant links with knet. `pcs_cluster` EXAMPLES contains example with 3 links to see the syntax. Change was done in commit https://github.com/OndrejHome/ansible.pcs-modules-2/commit/887386b48ea77959801d1b451f90786d6c5bbca7
(note that above change doesn't have tag (yet), so it is only in `master`)

Have a good start into this week.

**Comment by SpitchAG** [comment]
mm i am still confused about what to be done:

- (required, already present in PR) edit that adds RRP IP to pcs_cluster module 
not sure what it means because you also request to change rrp_ip in rrp_interface, so the current PR doesnt address that

- (required) implement check to verify that preferred RRP address is different from ring_0 address
if i use rrp_interface, rrp address cannot be the same (unless the 2nics register the same ip which doesnt seem to be a common case, or unless rrp_interface is also the primaty one)
-- > shall i just  check that rrp_interface is not the same as default_ipv4 interface?

thank you

**Comment by OndrejHome** [comment]
> (required, already present in PR) edit that adds RRP IP to pcs_cluster module
not sure what it means because you also request to change rrp_ip in rrp_interface, so the current PR doesnt address that

hmm, I think I wanted to say the we have already way for passing IP present here, but reading it again let me rephrase it:
- [ ] (required) pass the IP of `rrp_interface` to `pcs_cluster` modules `node_list`. Currently PR implements `rrp_ip` instead. Either 
- replace the `rrp_ip` with IP of `rrp_interface` or 
- (optional) make it into conditional that IP of `rrp_interface` is used only when `rrp_ip` is not present - this would allow your use case with IPs discussed earlier.

>  -- > shall i just check that rrp_interface is not the same as default_ipv4 interface?
I would go only with checking that 'IP' from `rrp_interface` is not same as `default_ipv4`.

Thank you for questions and sorry for delay.

**Comment by OndrejHome** [comment]
Hi @SpitchAG, It has been some days since I have heard from you. I want to check if there is anything you got stuck on or if your are just a bit busy these days. In any case please let me know so I can help if possible.

Ondrej

**Comment by SpitchAG** [comment]
Hi ! I just came back from holidays ans plan to work on this end if week or
next week,
Cheers

Le lun. 10 août 2020 à 14:37, Ondrej Faměra <notifications@github.com> a
écrit :

**Comment by SpitchAG** [comment]
Hi I pushed something,
let me know
tested on:
CentOS Linux release 7.7.1908
(updated kernel)

**Comment by OndrejHome** [comment]
Thank you for changes and update! I had a brief look at changes and looks so far good to me. Will give it some testing and unless there is something big I would merge it during  weekend.

**Comment by OndrejHome** [comment]
Hi @SpitchAG , I have tested the current version of PR on CentOS 7.8 and it works well in my test clusters.
I have tried combination fo only using rrp_interface, only rrp_ip and also trying wrong IPs.

Good job! Thank you for your contribution to this role!

Unless you are in hurry I plan to make new tag in master branch after taking care of another outstanding PR. If you would need a tagged version sooner, just let me know.

**Comment by SpitchAG** [comment]
thanks to you for this nice ansible role!
keep the good job & no, there is no urgency for us,
cheers

**Comment by OndrejHome** [comment]
Changes from this PR are now included in newly tagged version `25.0.0` that was also pushed to [Ansible Galaxy](https://galaxy.ansible.com/OndrejHome/ha-cluster-pacemaker).

**Comment by SpitchAG** [comment]
Thanks

Le sam. 5 sept. 2020 à 10:15, Ondrej Faměra <notifications@github.com> a
écrit :
